nixos/modules/services/system/swapspace.nix

```
@@ -0,0 +1,106 @@
+    package = mkPackageOption pkgs "swapspace" { };
+    settings = mkOption {
+      type = types.submodule {
+        options = {
+          swappath = mkOption {
+            type = types.str;
+            default = "/var/lib/swapspace";
+            description = "Location where swapspace may create and delete swapfiles";
+          };
+          lower_freelimit = mkOption {
+            type = types.ints.between 0 100;
+            default = 20;
+            description = "Lower free-space threshold: if the percentage of free space drops below this number, additional swapspace is allocated";
+          };
+          upper_freelimit = mkOption {
+            type = types.ints.between 0 100;
+            default = 60;
+            description = "Upper free-space threshold: if the percentage of free space exceeds this number, swapspace will attempt to free up swapspace";
+          };
+          freetarget = mkOption {
+            type = types.ints.between 0 100;
+            default = 30;
+            description = ''
+              Percentage of free space swapspace should aim for when adding swapspace.
+              This should fall somewhere between lower_freelimit and upper_freelimit.
+            '';
+          };
+          min_swapsize = mkOption {
+            type = types.str;
+            default = "4m";
+            description = "Smallest allowed size for individual swapfiles.";
+          };
+          max_swapsize = mkOption {
+            type = types.str;
+            default = "2t";
+            description = "Greatest allowed size for individual swapfiles.";
+          };
+          cooldown = mkOption {
+            type = types.ints.unsigned;
+            default = 600;
+            description = ''
+              Duration (roughly in seconds) of the moratorium on swap allocation that is instated if disk space runs out, or the cooldown time after a new swapfile is successfully allocated before swapspace will consider deallocating swap space again.
+              The default cooldown period is about 10 minutes.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Config file for swapspace.
+        See the options here: <https://github.com/Tookmund/Swapspace/blob/master/swapspace.conf>
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+    systemd.services.swapspace = {
+      after = [
+        "local-fs.target"
+        "swap.target"
+      ];
+      requires = [
+        "local-fs.target"
+        "swap.target"
+      ];
+      wantedBy = [ "multi-user.target" ];
+      documentation = [ "man:swapspace(8)" ];
+      description = "Swapspace, a dynamic swap space manager";
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = "${lib.getExe cfg.package} -c ${configFile}";
+        Restart = "always";
+        RestartSec = 30;
```

```suggestion
        ExecStart = "${lib.getExe cfg.package} -c ${configFile}";
```
This is already present in the [upstream systemd unit config](https://github.com/Tookmund/Swapspace/blob/9342d856ccc4c382c0c624e985d9c2b7b9fd9f57/swapspace.service), right? So can't we just remove this here?

---

I am unfamiliar with systemd and nix in general, but how does nix know about the upstream systemd config? I didn't add it anywhere in the service.

---

My mistake, I somehow thought that this was already done. You have to add `systemd.packages = [ cfg.package ];`, for example below `environment.systemPackages = [ cfg.package ];`.
When the `swapspace` package is built, the upstream unit config is copied to the output: https://github.com/NixOS/nixpkgs/blob/a0b1b922588da93b23b59d271570bb8068546b61/pkgs/tools/admin/swapspace/default.nix#L35
You can see this file, if you build the package with `nix-build -A swapspace` and then list all the files with `tree result`.
Then `systemd.packages = [ cfg.package ];` causes the file from the output to be symlinked to `/etc/systemd/system/swapspace.service`. The extra options set in the NixOS module are then written to a separate file in the Nix store and symlinked to `/etc/systemd/system/swapspace.service.d/overrides.conf`. systemd then merges the two files in memory when it reads them.

---

done